moveit_kinematics/kdl_kinematics_plugin/src/chainiksolver_pos_nr_jl_mimic.cpp

```diff
@@ -110,79 +110,81 @@ void ChainIkSolverPos_NR_JL_Mimic::qMimicToq(const JntArray& q, JntArray& q_resu
 int ChainIkSolverPos_NR_JL_Mimic::CartToJnt(const JntArray& q_init, const Frame& p_in, JntArray& q_out)
 {
   //  Note that q_init and q_out will be of size chain.getNrOfJoints()
-  //  qToqMimic(q_init,q_temp);
-
-  q_temp = q_init;
+  double last_delta_twist_err = DBL_MAX;
+  double step_size = 1.0;
+  q_out = q_init;
   ROS_DEBUG_STREAM_NAMED("kdl", "Input:");
   for (std::size_t i = 0; i < q_out.rows(); ++i)
-    ROS_DEBUG_NAMED("kdl", "%d: %f", (int)i, q_out(i));
+    ROS_DEBUG_NAMED("kdl", "%d: %f", (int)i, q_init(i));
 
   unsigned int i;
+  bool success = false;
   for (i = 0; i < maxiter; ++i)
   {
-    fksolver.JntToCart(q_temp, f);
+    fksolver.JntToCart(q_out, f);
     delta_twist = diff(f, p_in);
 
-    if (position_ik)
-    {
-      if (fabs(delta_twist(0)) < eps && fabs(delta_twist(1)) < eps && fabs(delta_twist(2)) < eps)
-        break;
-    }
-    else
-    {
-      if (Equal(delta_twist, Twist::Zero(), eps))
-        break;
+    // check norms of position and orientation errors
+    double position_error = delta_twist.vel.Norm();
+    double orientation_error = position_ik ? 0 : delta_twist.rot.Norm();
+    double delta_twist_err = std::max(position_error, orientation_error);
+    if (delta_twist_err <= eps) {
+      success = true;
+      break;
     }
 
-    ROS_DEBUG_STREAM_NAMED("kdl", "delta_twist");
-    for (std::size_t i = 0; i < 6; ++i)
-      ROS_DEBUG_NAMED("kdl", "%d: %f", (int)i, delta_twist(i));
-
-    iksolver.CartToJnt(q_temp, delta_twist, delta_q);
-
-    Add(q_temp, delta_q, q_temp);
+    if (delta_twist_err >= last_delta_twist_err) {
+      // if the error increased, we are close to a singularity -> reduce step size
+      double old_step_size = step_size;
+      step_size *= 0.5; // reduce scale;
```

Comment:
seems to me there should be a min step_size that should be checked s.t. the value can't get too close to zero. does that make sense here?

Reply:
There is a check a few lines down: `if (step_size < eps)`.